src/mcpwm/timer.rs

```diff
@@ -0,0 +1,302 @@
+use core::convert::TryInto;
+use core::marker::PhantomData;
+use core::ptr;
+
+use esp_idf_sys::{
+    esp, mcpwm_del_timer, mcpwm_new_timer, mcpwm_timer_config_t,
+    mcpwm_timer_config_t__bindgen_ty_1, mcpwm_timer_count_mode_t,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_DOWN,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_PAUSE,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP_DOWN, mcpwm_timer_enable,
+    mcpwm_timer_handle_t, mcpwm_timer_start_stop,
+    mcpwm_timer_start_stop_cmd_t_MCPWM_TIMER_START_NO_STOP,
+    soc_periph_mcpwm_timer_clk_src_t_MCPWM_TIMER_CLK_SRC_DEFAULT,
+};
+
+use crate::mcpwm::Group;
+use crate::prelude::{FromValueType, Hertz};
+
+use super::operator::NoOperator;
+use super::timer_connection::TimerConnection;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct TimerConfig {
+    resolution: Hertz,
+    period_ticks: u16,
+    count_mode: CountMode,
+    // TODO
+    // on_full: FF,
+    // on_empty: FE,
+    // on_stop: FS,
+}
+
+impl Default for TimerConfig {
+    fn default() -> Self {
+        Self {
+            resolution: 80.MHz().into(),
+            period_ticks: 8_000, // 10kHz
+            count_mode: CountMode::Up,
+        }
+    }
+}
+
+impl TimerConfig {
+    /*#[must_use]
+    pub fn resolution(mut self, resolution: impl Into<Hertz>) -> Self {
+        self.resolution = resolution.into();
+        self
+    }*/
+
+    // TODO: make sure this description is accurate
+    /// Set number of ticks per period
+    ///
+    /// This is inversely proportional to the frequency of the signal
+    ///
+    /// You can calculate the frequency as
+    /// `frequency = resolution / period_ticks`
+    ///
+    /// For example a resolution of 80MHz and a period_ticks of 8_000:
+    /// `10kHz = 80MHz / 8_000`
+    ///
+    /// NOTE: This will be the same as `Self::get_period_peak` for all `CounterMode` except for
+    /// `CounterMode::UpDown` where the period will be twice as large as the peak value since
+    /// the timer will count from zero to peak and then down to zero again
+    #[must_use]
+    pub fn period_ticks(mut self, period_ticks: u16) -> Self {
+        self.period_ticks = period_ticks;
+        self
+    }
+
+    #[must_use]
+    pub fn count_mode(mut self, counter_mode: CountMode) -> Self {
+        self.count_mode = counter_mode;
+        self
+    }
+}
+
+pub struct Timer<const N: u8, G: Group> {
```

Comment:
Having `_group` and `_timer` undescored means you don't actually use them after the `new` constructor. Can we get rid of those? That would mean getting rid of the `N` and `G` generics on the `Timer(Driver)` which simplifies the type.

Reply:
See [discussion](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1035028499). For(not added in this PR) to be able to support some limited version of this for IDF 4.4 later on, we need to be able to limit what `Timer` can be attached to what `Operator`.

In general a timer can only be connected to operators in the same `Group`, this is a hardware limitation. With IDF 4.4 there is also the constraint that `Timer<X, G>` can only be connect to `Operator<Y, G>` iff `X=Y`. This is done using (misspelled) attatch_operator{A} (soon to be `attach_operator{A}`).

Reply:
Anyway, I could not come up with a less generics heavy way to express something like that. I am however very open to suggestions :)

Reply:
If the `G` generic is remaining, that's still not the end of the world, as long as we can still simplify / remove most of the other generics induced by traits in the code.